docs/architecture/components/index.md

```diff
@@ -0,0 +1,171 @@
+        cg.add(var.set_baz(baz))
+```
+
+Let's break this down a bit.
+
+```python
+import esphome.config_validation as cv
+import esphome.codegen as cg
+```
+
+`config_validation` is a module that contains all the common validation functions that are used to validate the configuration.
+Components may contain their own validations as well and this is very extensible.
+`codegen` is a module that contains all the code generation functions that are used to generate the C++ code that is placed
+into `main.cpp`.
+
+
+```python
+example_component_ns = cg.esphome_ns.namespace("example_component")
+```
+
+This is the c++ namespace inside the `esphome` namespace. It is required here so that the codegen knows the exact 
+namespace of the class `that is being created. The namespace **must** match the name of the component.
+
+
+```python
+ExampleComponent = example_component_ns.class_("ExampleComponent", cg.Component)
+```
+
+This is the class that is being created. The first argument is the name of the class, the second and subsequent 
+arguments are the base classes that this class inherits from.
+
+```python
+CONFIG_SCHEMA = cv.Schema({
+    cv.GenerateID(): cv.declare_id(ExampleComponent),
+    cv.Required(CONF_FOO): cv.boolean,
+    cv.Optional(CONF_BAR): cv.string,
+    cv.Optional(CONF_BAZ): cv.int_range(0, 255),
+})
+```
+
+This is the schema that will allow user configuration for this component. This example shows that the user must provide
+a `foo` value that is a boolean. The user may optionally provide a `bar` value that is a string, and a `baz` value that is
+an integer between 0 and 255. The `cv.GenerateID()` is a special function that generates a unique ID for this component but
+also allows the user to specify an `id` in the configuration for automations.
+
+
+
+The `to_code` method is called after the entire configuration has been validated. This function is given the
+parsed `config` object for this instance of this component. This method is responsible for generating the
+C++ code that is placed into `main.cpp` translates the user configuration into the C++ instance method calls,
+setting the variables on the object.
+
+```python
+var = cg.new_Pvariable(config[CONF_ID])
+```
+
+`var` becomes a special object that represents the actual C++ object that will be generated. The `CONF_ID` that represents
+the above `cv.GenerateID()` holds both the id string, and the class name of the component `ExampleComponent`. Subsequent arguments
+to `new_Pvariable` are arguments that can be passed to the constructor of the class.
+
+```python
+await cg.register_component(var, config)
+```
+
+This line generates `App.register_component(var)` in C++ which registers the component so that its `setup`, `loop` and/or `update`
+functions are called correctly.
+
+```python 
+cg.add(var.set_foo(config[CONF_FOO]))
+```
+
+```c++
+var->set_foo(true);
+```
+
+This would become the above C++ assuming the user set `foo: true` in the YAML configuration.
```

Comment:
Just rearranging this ever-so-slightly. haha

````suggestion
Assuming the user has `foo: true` in their YAML configuration, this line:

```python 
cg.add(var.set_foo(config[CONF_FOO]))
```

...will result in this line:

```c++
var->set_foo(true);
```

...in the generated `main.cpp` file.
````